### PR TITLE
dsl context block helper

### PR DESCRIPTION
**Pull request by freels**
Provides a helper to build a query using the dsl in an unscoped fashion:

```ruby
include Fauna

query do 
  create Ref.new('classes/foo'), object(data: object(foo: 'bar'))
end
```

### CODE REVIEW

**Comment by freels** [comment]
First in a few dsl revision PRs...

File: lib/fauna/query.rb [hunk 1]
```diff
@@ -451,13 +457,13 @@ def self.or(*booleans)
     # A not function
     #
     # Reference: {FaunaDB Miscellaneous Functions}[https://faunadb.com/documentation/queries#misc_functions]
-    def self.not(boolean)
+    def not(boolean)
       { not: boolean }
     end
 
-  private
+    private
```
Comment (andy-faunadb) [inline lib/fauna/query.rb]: We've been intentionally dedenting `private` statements.

**Comment by freels** [comment]
THe build is actually passing, but I had an idea that will make this change a bit better.

File: lib/fauna/query.rb [hunk 2]
```diff
@@ -473,8 +479,49 @@ def self.with_auto_vars(n_vars)
     #
     # This ensures that a single value passed is not put in an array, so
     # <code>query.add([1, 2])</code> will work as well as <code>query.add(1, 2)</code>.
-    def self.varargs(values)
+    def varargs(values)
       values.length == 1 ? values[0] : values
     end
   end
+
+  def query(&block)
```
Comment (ashfire908) [inline lib/fauna/query.rb]: Is this intended to be used directly or by `Client.query` only?

File: lib/fauna/query.rb [hunk 3]
```diff
@@ -473,8 +479,49 @@ def self.with_auto_vars(n_vars)
...
+
+  ensure
+    dsl.instance_variables.each do |iv|
+      if iv.to_sym != :@__ctx__
+        ctx.instance_variable_set(iv, dsl.instance_variable_get(iv))
+      end
+    end
+  end
+
+  class DSLContext
```
Comment (andy-faunadb) [inline lib/fauna/query.rb]: This should be documented or have `:nodoc:`.

File: lib/fauna/client.rb [hunk 4]
```diff
@@ -104,10 +104,15 @@ def delete(path, data = {})
     # for information on constructing queries.
     #
     # +expression+:: A query expression
+    # +expr_block+:: May be provided instead of expression. Block is used to build an expression with Fauna.query.
```
Comment (ashfire908) [inline lib/fauna/client.rb]: Make sure docs link directly back to `Fauna.query`

Comment (andy-faunadb) [inline lib/fauna/client.rb]: This definitely needs a code example for both styles.

File: lib/fauna/query.rb [hunk 5]
```diff
@@ -473,8 +499,58 @@ def self.with_auto_vars(n_vars)
...
     # This ensures that a single value passed is not put in an array, so
     # <code>query.add([1, 2])</code> will work as well as <code>query.add(1, 2)</code>.
-    def self.varargs(values)
+    def varargs(values)
       values.length == 1 ? values[0] : values
     end
   end
+
+  ##
+  # Build a query expression.
```
Comment (ashfire908) [inline lib/fauna/query.rb]: Expand documentation with the requirements for using this and example code blocks. The end user of this should understand how to use it without assistance or having to reverse the code.

File: lib/fauna/query.rb [hunk 6]
```diff
@@ -473,8 +499,61 @@ def self.with_auto_vars(n_vars)
...
+  #
+  # Allows for unscoped calls to expression methods within the
```
Comment (andy-faunadb) [inline lib/fauna/query.rb]: "expression methods" should probably be "Fauna.Query methods"

File: lib/fauna/client.rb (hunk 4 above)
Comment (freels) [inline lib/fauna/client.rb]: done

File: lib/fauna/query.rb (hunk 5 above)
Comment (freels) [inline lib/fauna/query.rb]: done

**Comment by ashfire908** [comment]
Should/can we promote this via the README as something to use?

**Comment by freels** [comment]
Yeah, I can revise the README as part of final pass.

File: test/client_test.rb [hunk 7]
```diff
@@ -68,6 +68,11 @@ def test_ping
     assert_equal 'Scope All is OK', client.ping(scope: 'all')
   end
 
+  def test_query
+    assert (client.query { paginate(Ref.new('classes')) })[:data].is_a?(Array)
+    assert client.query(Fauna.query { paginate(Ref.new('classes')) })[:data].is_a?(Array)
```
Comment (ashfire908) [inline test/client_test.rb]: I'd like to see the non `Fauna.query` form tested as well, and that the output of all three are the same.

Comment (freels) [inline test/client_test.rb]: sure

File: lib/fauna/query.rb [hunk 8]
```diff
@@ -1,4 +1,7 @@
 module Fauna
+
+  extend self
```
Comment (andy-faunadb) [inline lib/fauna/query.rb]: Why is this one necessary?

File: lib/fauna/client.rb [hunk 9]
```diff
@@ -104,10 +104,23 @@ def delete(path, data = {})
...
+    # <code>client.query(Fauna::Query.add(1, 2, Fauna::Query.subtract(3, 2)))</code>
+    #
+    # Example using block:
+    #
+    # <code>client.query { add(1, 2, subtract(3, 2)) }</code>
     #
     # :category: Query Methods
-    def query(expression)
-      post('', expression)
+    def query(expression = nil, &expr_block)
```
Comment (andy-faunadb) [inline lib/fauna/client.rb]: This is being documented under "Miscellaneous Functions", making it look like it's part of the Query module. Maybe this code should be moved to a different file.

File: lib/fauna/query.rb (hunk 8 above)
Comment (freels) [inline lib/fauna/query.rb]: So you can call `Fauna.query`, and if you include Fauna, unscoped `query` also works.

Comment (freels) [inline lib/fauna/query.rb]: Arguably, the latter is not necessary, since client.query is more useful, will probably be more used.

File: lib/fauna/client.rb (hunk 9 above)
Comment (freels) [inline lib/fauna/client.rb]: It's an instance method on client, so I'm not sure if moving it makes sense. Is rdoc doing the wrong thing here? If it were moved up, would it be more prominent in rdoc?

File: test/query_test.rb [hunk 10]
```diff
@@ -1,70 +1,77 @@
...
     @ref_m1 = create_instance(m: 1)[:ref]
     @ref_n1m1 = create_instance(n: 1, m: 1)[:ref]
 
-    @thimble_class_ref = client.post('classes', name: 'thimbles')[:ref]
+    client.post('classes', name: 'thimbles')
+  end
+
+  def test_query_helper_method_missing
+    foo = "foo"
+    assert_equal "foo", Fauna.query { foo }
```
Comment (ashfire908) [inline test/query_test.rb]: `assert_equal foo, Fauna.query { foo }`? Or is there a reason not to do that

File: test/query_test.rb [hunk 11]
```diff
@@ -1,70 +1,77 @@
...
   end
 
   def test_object
-    # unlike quote, contents are evaluated
-    assert_query(
-      { x: 1 },
-      Query.object(x: Query.let({ x: 1 }, Query.var(:x))))
-  end
-
-  def test_quote
```
Comment (ashfire908) [inline test/query_test.rb]: Why was this removed?

Comment (freels) [inline test/query_test.rb]: quote is going away

File: test/query_test.rb (hunk 10 above)
Comment (freels) [inline test/query_test.rb]: probably doesn't make a difference.

File: lib/fauna/query.rb [hunk 12]
```diff
@@ -41,23 +62,31 @@ def self.var(name)
     # An if expression
     #
     # Reference: {FaunaDB Basic Forms}[https://faunadb.com/documentation/queries#basic_forms]
-    def self.if(condition, then_, else_)
+    def if(condition, then_, else_)
```
Comment (ashfire908) [inline lib/fauna/query.rb]: Do we maybe want to just replace these all with their non-keyword replacements?

File: test/query_test.rb [hunk 13]
```diff
@@ -343,137 +329,109 @@ def test_time
...
-    assert_query 2, Query.subtract(2)
-    assert_bad_query Query.subtract
+    assert_equal -6, client.query { subtract(2, 3, 5) }
+    assert_equal 2, client.query { subtract(2) }
   end
 
   def test_divide
-    assert_query 2.0 / 15, Query.divide(2.0, 3, 5)
-    assert_query 2, Query.divide(2)
-    assert_bad_query Query.divide(1, 0)
```
Comment (ashfire908) [inline test/query_test.rb]: Do we no longer want to assert against known bad queries?

Comment (freels) [inline test/query_test.rb]: yes. assuming the error handling mechanism is tested, behavioral tests of individual functions are out of scope for these.

File: lib/fauna/query.rb (hunk 12 above)
Comment (freels) [inline lib/fauna/query.rb]: I guess so. :(

File: .rubocop.yml [hunk 14]
```diff
@@ -62,9 +65,15 @@ Style/EachWithObject:
 Style/Encoding:
   Enabled: false
 
+Style/EmptyLinesAroundClassBody:
```
Comment (ashfire908) [inline .rubocop.yml]: Don't see the harm in keeping this on.

File: lib/fauna/query.rb [hunk 15]
```diff
@@ -88,7 +108,7 @@ def self.quote(expr)
     #   Takes one or more ::var expressions and uses them to construct an expression.
     #   If this takes more than one argument, the lambda destructures an array argument.
     #   (To destructure single-element arrays use ::lambda_expr.)
-    def self.lambda(&block)
+    def lambda(&block)
```
Comment (andy-faunadb) [inline lib/fauna/query.rb]: This shares a name with a builtin, is that a problem?

File: .rubocop.yml (hunk 14 above)
Comment (freels) [inline .rubocop.yml]: I disabled it because it was causing warnings that were annoying.

File: lib/fauna/query.rb (hunk 15 above)
Comment (freels) [inline lib/fauna/query.rb]: Surprisingly, no.

File: .rubocop.yml (hunk 14 above)
Comment (ashfire908) [inline .rubocop.yml]: When I disabled this and ran it, all it did was raise a single warning about some new line spacing. I think keeping it on will enforce some style consistency.

Comment (freels) [inline .rubocop.yml]: Yeah, I know. But the case it complained about was wrong IMO, which was why I disabled it.

**Comment by ashfire908** [comment]
LGTM

**Comment by ashfire908** [comment]
Actually, the README needs to be updated still to demo this

File: lib/fauna/query.rb [hunk 16]
```diff
@@ -8,14 +8,34 @@ module Fauna
   # Example:
   #
   #   query = Fauna::Query.create(Fauna::Ref.new('classes', 'spells'), Fauna::Query.quote(data: { name: 'Magic Missile' }))
+
+  ##
```
Comment (ashfire908) [inline lib/fauna/query.rb]: You need to place this above the doc above, which is the doc for the `Fauna::Query` module.

Comment (freels) [inline lib/fauna/query.rb]: I don't understand. This is the doc for the method.

**Comment by ashfire908** [comment]
:sheep:
